packages/core/src/routes/sign-in-experience/custom-ui-assets/index.ts

```diff
@@ -0,0 +1,100 @@
+import { readFile } from 'node:fs/promises';
+
+import { uploadFileGuard, maxUploadFileSize } from '@logto/schemas';
+import { generateStandardId } from '@logto/shared';
+import { object, z } from 'zod';
+
+import { EnvSet } from '#src/env-set/index.js';
+import RequestError from '#src/errors/RequestError/index.js';
+import koaGuard from '#src/middleware/koa-guard.js';
+import SystemContext from '#src/tenants/SystemContext.js';
+import assertThat from '#src/utils/assert-that.js';
+import { getConsoleLogFromContext } from '#src/utils/console.js';
+import { buildAzureStorage } from '#src/utils/storage/azure-storage.js';
+import { getTenantId } from '#src/utils/tenant.js';
+
+import { type ManagementApiRouter, type RouterInitArgs } from '../../types.js';
+
+export default function customUiAssetsRoutes<T extends ManagementApiRouter>(
+  ...[router]: RouterInitArgs<T>
+) {
+  // TODO: Remove
+  if (!EnvSet.values.isDevFeaturesEnabled) {
+    return;
+  }
+
+  router.post(
+    '/sign-in-exp/default/custom-ui-assets',
+    koaGuard({
+      files: object({
+        file: uploadFileGuard.array().min(1),
+      }),
+      response: z.object({
+        customUiAssetId: z.string(),
+      }),
+      status: [200, 400, 500],
+    }),
+    async (ctx, next) => {
+      const { file: bodyFiles } = ctx.guard.files;
+      const file = bodyFiles[0];
+
+      assertThat(file, 'guard.invalid_input');
+      assertThat(file.size <= maxUploadFileSize, 'guard.file_size_exceeded');
+      assertThat(file.mimetype === 'application/zip', 'guard.mime_type_not_allowed');
+
+      const { experienceZipsProviderConfig } = SystemContext.shared;
+      assertThat(
+        experienceZipsProviderConfig?.provider === 'AzureStorage',
+        'storage.not_configured'
+      );
+      const { connectionString, container } = experienceZipsProviderConfig;
+
+      const { uploadFile, downloadFile, isFileExisted } = buildAzureStorage(
+        connectionString,
+        container
+      );
+
+      const [tenantId] = await getTenantId(ctx.URL);
+      assertThat(tenantId, 'guard.can_not_get_tenant_id');
+
+      const customUiAssetId = generateStandardId(8);
+      const objectKey = `${tenantId}/${customUiAssetId}/${file.originalFilename}`;
```

Comment:
will it be better to normalize the file name?

Reply:
Good to me. Will it be OK to name it something like `${tenantId}/${customUiAssetId}/assets.zip`?